src/SharedStream.mo

```diff
@@ -31,25 +37,29 @@ module {
   /// };
   public class StreamReceiver<T>(
     streamId : Nat,
-    callback : (streamId : Nat, item : T, index : Nat) -> (),
     startFromIndex : Nat,
+    closeStreamTimeoutSeconds : Nat,
+    itemCallback : (streamId : Nat, item : T, index : Nat) -> (),
+    chunkErrorCallback : (expectedIndex : Nat, receivedIndex : Nat) -> (),
   ) {
 
     var expectedNextIndex_ : Nat = startFromIndex;
-
     var lastChunkTimestamp : Time.Time = Time.now();
 
-    public func isStreamClosed() : Bool = (Time.now() - lastChunkTimestamp) > 120_000_000_000; // 2 minutes
+    let timeout = closeStreamTimeoutSeconds * 1_000_000_000;
+
+    public func isStreamClosed() : Bool = (Time.now() - lastChunkTimestamp) > timeout;
 
-    public func onChunk(chunk : [T], firstIndex : Nat) : R.Result<(), { #StreamClosed }> {
+    public func onChunk(chunk : [T], firstIndex : Nat) : R.Result<(), ChunkError> {
       if (isStreamClosed()) {
-        return #err(#StreamClosed);
+        return #err(#StreamClosed(expectedNextIndex_));
       };
       if (firstIndex != expectedNextIndex_) {
-        Debug.trap("Broken chunk index: " # Nat.toText(firstIndex) # "; expected: " # Nat.toText(expectedNextIndex_));
+        chunkErrorCallback(expectedNextIndex_, firstIndex);
```

Comment:
Sorry, I must have been confused when I wrote about the callback in error case (first index != expected index). Now that we return the #BrokenPipe error we don't need the ErrorCallback anymore. We can remove the error callback.

Reply:
removed